arch: arm: fix K_MEM_PARTITION_IS_WRITABLE definition

ioannisg opened the pull request:
In ARMv8-M MPU it is not possible to have the following access
permissions: Privileged RW / Unprivileged RO. So we define
K_MEM_PARTITION_IS_WRITABLE macro separately for v8M and v7M MPU
architectures (in the separate include files).

The commit does not change the behavior for ARMv6-M and ARMv7-M. It only corrects the implementation for ARMv8-M.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: fix K_MEM_PARTITION_IS_WRITABLE definition

In ARMv8-M MPU it is not possible to have the following access
permissions: Privileged RW / Unprivileged RO. So we define
K_MEM_PARTITION_IS_WRITABLE macro separately for v8M and v7M MPU
architectures (in the separate include files).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
FYI, @pizi-nordic and @agross-linaro 
>In ARMv8-M MPU it is not possible to have the following access
permissions: Privileged RW / Unprivileged RO.

Please check this in the ARMv8-M Architecture Reference Manual, page D-1063:
"D1.2.150 MPU_RBAR, MPU Region Base Address Register"

ioannisg commented:
@agross-linaro could you take a quick look?

galak reviewed (changes requested):
Do we need to fixup samples/mpu/mem_domain_apis_test/src/main.c & tests/kernel/mem_protect/mem_protect/src/mem_domain.c for armv8m?

ioannisg commented:
>Do we need to fixup samples/mpu/mem_domain_apis_test/src/main.c & tests/kernel/mem_protect/mem_protect/src/mem_domain.c for armv8m?

Yes, I think so.

ioannisg commented:
@galak those samples won't compile for ARMv8m as is. However, we still have no v8m board available in the tree.

galak commented:
> @galak those samples won't compile for ARMv8m as is. However, we still have no v8m board available in the tree.

Will they break when we have an in tree board and enable the support?  Can we at least put some handling for ARMv8m in those tests/samples?

ioannisg commented:
> 
> 
> > @galak those samples won't compile for ARMv8m as is. However, we still have no v8m board available in the tree.
> 
> Will they break when we have an in tree board and enable the support? Can we at least put some handling for ARMv8m in those tests/samples?

Yes, @galak , I can do that. I will push the patches in this PR and ping you back.

agross-oss reviewed (approved):
this looks good to me

ioannisg committed:
arch: arm: fix mem domain sample/test for ARMv8-M access permissions

This commit updates the mem_domain_apis_test sample and the
mem_protect test, so they can compile and execute in ARMv8-M
platforms, which do not support the P_RW_U_RO access permissions
combination (privileged read/write, unprivileged read-only). The
modification consists of, simply, selecting a different access
permission (P_RO_U_RO) when building for ARMv8-M MPUs with the
unmodified ARM MPU architecture.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@galak , I pushed a second commit to select special access permissions for v8m MPU (as we already have for x86, actually) 
Please, re-review.
